syntax = "proto3";
import "google/protobuf/timestamp.proto";



package DSM;

option go_package = "github.com/EricChiquitoG/Remanet_DSM/DSM_protos";

service SubmissionService {
    rpc CheckAvailabilty(Process) returns (ProcessResponse) {}
    rpc CheckInterest(Purchase) returns (PurchaseResponse) {}
    rpc EnrollServer(Enroll) returns (EnrollResponse) {}
    rpc Optimize (OptimizationRequest) returns (OptimizationResponse) {}
}

message Process {
    string step_name = 1;            // Name of the step
    string product_type = 2;
    string economic_operator = 3;   // Name or identifier of the economic operator
    google.protobuf.Timestamp submitted_at = 4;        // Submission timestamp (ISO 8601 format recommended)
    repeated string requirements = 5; // List of requirements as strings
}

message ProcessResponse {
    string status = 1;        // Status of the submission (e.g., "success" or "failure")
    string message = 2;       // Additional details or error message
    repeated string capability = 3;
}

message Purchase {
    string product_type = 1;
    string amount = 2;
    repeated float location = 3;
    double logistics = 4;
    double co2 = 5;
    double energy = 6;
    double costEUR = 7;
    repeated string requirements = 8;
}

message PurchaseResponse {
    string status = 1;
    string message = 2;
    bool capability = 3;

}

message Enroll {
    string Name = 1;
    string Address = 2;
    repeated double Location = 3;
    double CostH = 4;
    repeated string Offerings = 5;
}

message EnrollResponse {
    string status = 1;
    string message = 2;
}

message OptimizationRequest {
    // Structure should match the data needed by your Python optimization code
    // Based on your previous discussion, this might include:
    // Or you could structure it more granularly:
    string json_problem_data = 1; // Send the entire problem JSON as a string
}
  
  
  // Message for the response (data sent from Python back to Go)
  message OptimizationResponse {
    // Structure should match the results you want to send back
    // This could be the Pareto front solutions and objectives
    repeated Solution solutions = 1;
    string error_message = 2; // Optional field for error handling
  }
  
  // Message to represent a single solution from the optimization
  message Solution {
      repeated int32 user_indices = 1; // The sequence of user indices
      repeated string user_ids = 2; // The sequence of user IDs
      double transport_cost = 3;
      double manufacturing_cost = 4;
      // Add other relevant solution details if needed
  }
  